### PR TITLE
Work together?

### PR DESCRIPTION
**Pull request by til-schneider**
I recently forked your great project ansel.

Here's my motivation: I used Google Picasa for a while, but Google announced not to update Picasa any more - they want the people to upload their data... So I searched for an alternative, but didn't find something suitable. So I decided to developing something. Then I found your project which goes pretty much in the direction I wanted to go (web based using electron).

This is what I've done so far:

- Made project work on Mac OS
- Added TypeScript support
- Removed Linux USB support from master branch into a new branch linux-usb
- Showing PictureDetail in full size
- Added a UI sandbox for developing React components in a sandboxed way
- Fixed some bugs

This is what I'd like to add:

- Simple editing of images (rotating in 90° steps, cutting, saturation, brightness, ...)
- Reading Picasa settings
- Write own settings like Picasa did:
  - Don't change original image
  - Remember the actions in a hidden file per folder
  - Apply actions on export
- Change to UI theme into neutral colors (similar to Adobe Lightroom)

What is your status about the project? Are you planning to work on it?

If you like the direction I'd like to go, I would be happy if we could work together in a common repo.

### CODE REVIEW

**Comment by m0g** [comment]
Wow, that's really impressive.

To be honest, I was expecting someone to get interested into this project. And I also haven't worked on it for a while.

I can see that you've made a lot changes, some more opinionated than others.

But of course I'm welcoming your work and I'm really happy about your contribution.

A couple of remarks:

* **Typescript**: I believe it's a great addition to the project and also quite necessary for such a big project
* **Exif-parser instead of exif2**: I'm not familiar with this project, maybe it did not yet exist when I was working on ansel. Seems to be working fine from my end. Plus no more native library (win win).
* **Picture detail in full screen**: I don't know if you noticed, but you could already do that in the past by simply pressing `<Tab>`
* **UI sandbox**: I'm not sure to understand what's going on in there.
* **GLFX canvas**: Could you explain me what this is for?


Yes now that I see that the project is gaining traction, I'd be interested to work more on it. If you want I could create a slack or a telegram group for the project.

As for what you'd like to work on:

* **Simple editing of image**: I think this depends a lot to whether people use raw or not. But yes we could do it in non destructive way, and either store changes in the DB or in the filesystem.
* **Reading Picasa settings:** I'm not familiar with Picasa, so I don't know what you want to do here.
* **Write own settings like Picasa did**: yes, could be a hidden file, but could also be stored in the DB. I believe lightroom uses the DB solution. I don't know what is the best approach.
* **Neutral theme**: I think on the long run we should support themes and let the user choose. 

Things I'd like to work on:

* Update redux implementation
* Move to CSS in JS
* Get rid of gulp in favor of npm scripts
* Get rid of bookshelf.js & knex.js and simply write raw queries
* Introducing the usage of async/await instead of promises

I'm also curious to hear what you use Ansel for? Is it for managing your smartphone pictures or for a DSLR/hybrid?

**Comment by til-schneider** [comment]
**opinionated changes**

> I can see that you've made a lot changes, some more opinionated than others.

Hmm, opinionated. You probably think of things like changing lint rules, omitting semicolons or changing indentation to 4 spaces. Yes, that's all personal preference. Sorry about that. When I first started, I didn't really think about pull requests and keeping diffs small. I'll try to avoid this in future. Please tell me, if you don't like some of the stuff I did (or will do).

**removing exiv2**

> Exif-parser instead of exif2: I'm not familiar with this project, maybe it did not yet exist when I was working on ansel. Seems to be working fine from my end. Plus no more native library (win win).

The reason for removing exiv2 was that it didn't work on Max OS.

I got the following error:
~~~
In file included from ../vendor/exiv2-0.25/src/canonmn.cpp:39:
../vendor/exiv2-0.25/src/i18n.h:35:11: fatal error: 'libintl.h' file not found
# include <libintl.h>
          ^~~~~~~~~~~
~~~

So this was more a part of making Ansel work on Mac OS. But I also think it's better to have a platform independent solution for this.

**Picture detail in full screen**

> I don't know if you noticed, but you could already do that in the past by simply pressing <Tab>

I didn't notice this hotkey. There are some more features I stumbled over when reading the code of which I don't know how to activate them - like opening an editor or showing a diff or what versions are about. Maybe be could add a visual way to activate those features. Hotkeys are a great thing for power users. But I think, there should always be a visual way, too.

In spite of that, I would prefer to keep the picture detail in full screen. One of my next steps would be to add simple image editing, for which I need a sidebar in the details. Having two sidebars would let few space for the actual image. I think in the details, the global filtering is not so much important and adds too much distraction.

**UI sandbox**

> UI sandbox: I'm not sure to understand what's going on in there.

This is just a little test UI where you can test React components in certain states. This is faster as navigating to the component and creating that state by hand.

**UI sandbox**

> GLFX canvas: Could you explain me what this is for?

This is a preparation for image editing. It shows the image in a WebGL canvas, which makes it possible to apply hardware accelerated effects (like saturation, brightness, sepia, blur, sharpen). Just have a look at the [demo page](http://evanw.github.io/glfx.js/demo/).

So the user can add or change effects and see the result instantly. Further more, the effects can be applied in runtime when viewing the images, so the effected image doesn't have to be stored on disk. It only needs the original image and a list of effects to apply.

**Simple editing of image**

> I think this depends a lot to whether people use raw or not. But yes we could do it in non destructive way, and either store changes in the DB or in the filesystem.

Yes, you're right. The solution I am going for would work on RGB data, not on raw. So this will be no Lightroom. Supporting raw would be much more effort, I think.

**Reading Picasa settings**

> I'm not familiar with Picasa, so I don't know what you want to do here.

Picasa store its changes to images in a `.picasa.ini` file in every image directory.

The `.picasa.ini` file contains stuff like that:

```
[Contacts2]
e7e10d12342c4b62=Mr Foo;;
[DSC08499.JPG]
rotate=rotate(3)
crop=rect64(3e0950bf7d0dc05f)
filters=crop64=1,3e0950bf7d0dc05f;
star=yes
faces=rect64(5d0d7be274239585),e7e10d12342c4b62
[DSC08381.JPG]
star=yes
```

So `DSC08499.JPG` has three effects, a star and a rectangle of a person's face (Mr Foo).

> Write own settings like Picasa did: yes, could be a hidden file, but could also be stored in the DB. I believe lightroom uses the DB solution. I don't know what is the best approach.

I would prefer writing similar files like Picasa. This keeps things much more open and it keeps the data near to the images. I use owncloud to share my image folders with my wife. The file based approach synchronizes the meta data, too. So we can collaborate. This approach also prevents data loss when uninstalling Ansel or when moving data to another computer. It even supports version control.

Therefore I think, the database should more act like a cache. But the actual data should always be kept in the file system from where Ansel syncs its internal state.

**Neutral theme**

> I think on the long run we should support themes and let the user choose.

Good idea. This should not be too complicated when using different sets of less variables.

**As for what you'd like to work on**

- **Update redux implementation:** Yes, great.
- **Move to CSS in JS:** As you may have noticed, I added little less files next to each styled component. This bloats up the number of files, of course. Maybe a CSS-in-JS approach would be nicer. Is the CSS stored in string? Don't you loose code completion in the IDE?
- **Get rid of gulp in favor of npm scripts:** Another possibility would be to change to webpack. This would allow creating a combined file (which speeds up start time), importing non-JavaScript assets and hot module replacement (which speeds up development a lot).
- **Get rid of bookshelf.js & knex.js and simply write raw queries:** Definitely a good idea.
- **Introducing the usage of async/await instead of promises:** Very nice.

**What I use Ansel for**

> I'm also curious to hear what you use Ansel for? Is it for managing your smartphone pictures or for a DSLR/hybrid?

Basically I want to manage my picture library, coming from smartphone and from my DSLR (currently  a DMC-G6). As I mentioned I currently manage my pics using Picasa in collaboration with my wife using owncloud synchronization. Which works pretty well.

**Comment by LeonardoGentile** [comment]
I just want to share my happiness for the direction this project is taking :]

It seems every little small thing that was possible to do back then 10 or 20 years ago on our old low powerful computers cannot be done today on our much more powerful machines and everything should be uploaded "in the cloud" as if our machines became suddenly dumb. 

I really like the personal cloud solutions like owncloud and keep things simple with folder+db file structure and avoid any lock-in solution.

I might be able to help in my spare time if we can organise

**Comment by m0g** [comment]
Great to hear that. Do you guys use telegram? We could move the conversation over there.

**Comment by til-schneider** [comment]
Telegram is OK for me (I sent you a mail with my number).

But for discussion about certain features, we could create issues here on github. This would make the discussion public and we would have them separated (one "thread" per feature, instead of a single Telegram group).

**Comment by m0g** [comment]
I agree, telegram is more for informal discussion. Other than that, tasks should be tracked by issues, with their respective pull requests

**Comment by til-schneider** [comment]
I opened issue #9 for discussing the "Simple editing of images" feature.

**Comment by til-schneider** [comment]
I just pushed an update which fixes anti aliasing
